### DISCUSSION TITLE
NodeJS chown error: EPERM: operation not permitted

### DISCUSSION THREAD

**Discussion by robertsLando** [community/community#26717] [Actions]
<p>I keep getting this error while running ci tests. In my tests I’m simply using <code>fs/promises</code> <code>chown</code> command to change ownership of a file placed in <code>/tmp/</code> dir.</p>
<p>I also tried using execSync from ‘child-process’ to execute the command and the error is the same so it’s not something related to fs module.</p>
<p>How could I fix this?</p>
<p>Test workflow that reproduces the error:</p>
<pre><code class="lang-yaml">name: test

on:
  pull_request:

jobs:
  test:
    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [14.x]
    
    steps:
      - name: Test
        run: |
          node -p "const fs=require('fs'); fs.writeFileSync('/tmp/test.txt', 'test'); fs.chownSync('/tmp/test.txt', 1000, 1000)"
</code></pre>
<p>Error:</p>
<pre><code class="lang-auto">Error: EPERM: operation not permitted, chown '/tmp/test.txt'
</code></pre>
<p>Running the same script with <code>sudo</code> it’s working.</p>
<p>Seems the problem is related to the user in virtual environment having uid 1001 and guid 121.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="robertsLando" data-post="1" data-topic="246151">
<div class="title">
<div class="quote-controls"></div>
 robertsLando:</div>
<blockquote>
<p>Running the same script with <code>sudo</code> it’s working.</p>
</blockquote>
</aside>
<p>That’s because a regular user generally can’t change the owner of files, they can create files with themselves as the owner where they can write. I think the real question is: Why do you <em>want</em> to change the ownership?</p>
<p>In the example I can’t see a reason, but I assume you simplified something.</p>

**Comment by robertsLando** [comment]
<p>The workflow I used here is just to show the problem, the reason is I got this problem while writing some test suites for a nodejs application. One of this tests creates some files and changes the ownership of them and that was failing.</p>
<p>The only workaround I found is to run test using sudo but I would like to know if there could be a better way to do it</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="robertsLando" data-post="3" data-topic="246151">
<div class="title">
<div class="quote-controls"></div>
 robertsLando:</div>
<blockquote>
<p>One of this tests creates some files and changes the ownership of them and that was failing.</p>
</blockquote>
</aside>
<p>If changing the ownership is relevant to the functioning of your application the application will need elevated privileges, and so will the tests for that functionality. Note that it doesn’t need full root privileges, the <code>CAP_CHOWN</code> capability should be sufficient. You could also run tests in a container so a random user running the tests won’t need admin access on the host system.</p>
<p>If it’s not actually relevant to the tests, they should try to do it.</p>
